### ISSUE TITLE
[FEATURE] Generate JPA boilerplate code for composite identifiers with @IdClass.

### ISSUE THREAD

**Issue by fabianMendez** [rzwitserloot/lombok#2032]
**Describe the feature**

Annotation to generate JPA boilerplate code for composite identifiers with @IdClass.

As stated [here](http://docs.jboss.org/hibernate/orm/5.4/userguide/html_single/Hibernate_User_Guide.html#identifiers-composite):

> * The composite identifier must be represented by a "primary key class". The primary key class may be defined using the javax.persistence.IdClass annotation.
> * The primary key class must be public and must have a public no-arg constructor.
> * The primary key class must be serializable.
> * The primary key class must define equals and hashCode methods.

Example:
```java
@Entity
@IdClassAnnotation
public class SystemUser {
	@Id
	private String subsystem;
	
	@Id
	private String username;
}
```

Will generate:
```java
@Entity
@IdClass(SystemUser.PK.class)
public class SystemUser {
	@Id
	private String subsystem;

	@Id
	private String username;

	public PK getId() {
		return new PK(subsystem, username);
	}

	public void setId(PK id) {
		this.subsystem = id.getSubsystem();
		this.username = id.getUsername();
	}

	public static class PK implements Serializable {
		private static final long serialVersionUID = 1L;

		private String subsystem;
		private String username;

		public PK() {
		}

		public PK(String subsystem, String username) {
			this.subsystem = subsystem;
			this.username = username;
		}

		public String getSubsystem() {
			return subsystem;
		}

		public void setSubsystem(String subsystem) {
			this.subsystem = subsystem;
		}

		public String getUsername() {
			return username;
		}

		public void setUsername(String username) {
			this.username = username;
		}

		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + ((subsystem == null) ? 0 : subsystem.hashCode());
			result = prime * result + ((username == null) ? 0 : username.hashCode());
			return result;
		}

		@Override
		public boolean equals(Object o) {
			if (o == this)
				return true;
			if (!(o instanceof PK))
				return false;
			PK other = (PK) o;
			if (!other.canEqual((Object) this))
				return false;
			if (!super.equals(o))
				return false;
			if (this.subsystem != other.subsystem)
				return false;
			if (this.username != other.username)
				return false;
			return true;
		}

		protected boolean canEqual(Object other) {
			return other instanceof PK;
		}
	}
}
```

**Describe the target audience**
Developers using the java persistence api, for instance spring developers using [spring data jpa](https://spring.io/projects/spring-data-jpa)

**Additional context**
1. I though this was not generic enough to be in this project but then I read [this](https://groups.google.com/forum/#!searchin/project-lombok/dataall%7Csort:date/project-lombok/RwdK8eCGnFc/Iy39rTISAAAJ) discussion and looks like this could go in the `lombok.extern.jpa` package

**Comment by rzwitserloot** [comment]
Sounds good. Am I understanding correctly that you're proposing the creation of both `lombok.extern.jpa.IdClassAnnotation` as well as `lombok.extern.jpa.Id`? If not, what are those `@Id` annotations on the fields (what is their fully qualified name), am I understanding correctly that lombok should surmise that the field(s) to be used as PK are the fields so annotated?

It's low priority; I'm not sure we're going to get to this anytime soon. However, a Pull Request would be accepted :)

**Comment by alvint** [comment]
Also, regardless of wether to produce similar hash codes, since `@IdClassAnnotation` implies that the class is an entity, and that implies even more boilerplate (at least the entity needs `equals` and `hashCode` as mentioned above), It may be simpler to instead of  `@IdClassAnnotation` just have a more general `@LombokEntity` annotation. This would automatically add `equals` and `hashCode` to the entity based on any `@Id` annotations, and if more than one `@Id` is present, create `PK`.

**Comment by Maaartinus** [comment]
I agree that generating `equals` and `hashCode` using the primary key is the right thing, but AFAIK most people use an "all-args-equals" instead. So this feature should be IMHO optional. Maybe there should be `@LombokEntity` with both features optional.

One tiny simplification: `SystemUser.hashCode` can delegate to `SystemUser.PK.hashCode`.

I'm unsure about the `null` treatment. For `equals`, this sounds good, but no matter how you define `hashCode`, it will most probably change when persisted. I myself let both method throw on nulls; that's probably too bold and incorrect, but so far it worked well.

**Comment by avirlrma** [comment]
Can I take this up?

**Comment by fabianMendez** [comment]
> 
> 
> Can I take this up?

Of course

**Comment by maxxyme** [comment]
any news...?
